// appwright/dist/vision/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VisionProvider = void 0;
const vision_1 = require("@empiricalrun/llm/vision");
class VisionProvider {
    device;
    webDriverClient;
    constructor(device, webDriverClient) {
        this.device = device;
        this.webDriverClient = webDriverClient;
    }
    async extractText(prompt) {
        const base64Screenshot = await this.webDriverClient.takeScreenshot();
        return await (0, vision_1.query)(base64Screenshot, prompt);
    }
    async tap(prompt) {
        const base64Screenshot = await this.webDriverClient.takeScreenshot();
        const { center, container: imageSize } = await (0, vision_1.getBoundingBox)(base64Screenshot, prompt);
        const driverSize = await this.webDriverClient.getWindowRect();
        const scaleFactorWidth = imageSize.width / driverSize.width;
        const scaleFactorHeight = imageSize.height / driverSize.height;
        if (scaleFactorWidth !== scaleFactorHeight) {
            console.warn(`Scale factors are different: ${scaleFactorWidth} vs ${scaleFactorHeight}`);
        }
        await this.device.tap({
            x: center.x / scaleFactorWidth,
            y: center.y / scaleFactorWidth,
        });
    }
}
exports.VisionProvider = VisionProvider;
